### PR TITLE
Refactors Scores and add querying/sorting options

### PR DESCRIPTION
**Pull request by florian-huber**
Here I changed quite some things to make the API nicer to work with:
- address #143 --> make `Scores.scores` a structured array (only when score contains more than one component, e.g. score + matches)
- add sorting option to ``scores_by_reference`` and ``scores_by_query`` method (implemented via ``BaseSimilarity``)
- fix bug in ``Spectrum()``: getter for metadata entries did not use ``copy()``
- fix bug/issue in ``Spectrum()``: ``__eq__`` method compared metadata dictionaries simply with ``==``. However, that fails when numpy array are added to the metadata (which we do in the ``add_fingerprints`` filter).
- added tests for those cases

In #152 we had also discussed adding ``Scores.top_scores_by_query(query, n=np.Inf)`` and ``Scores.top_scores_by_reference(reference, n=np.Inf)``, but this is not done here yet. For time reasons I would postpone it for another round of additions.

### CODE REVIEW

**Comment by florian-huber** [comment]
@sverhoeven Sorry for the back and forth (draft-not-draft...). Took me some time to find the bugs/issues that were causing weird behavior (they were in Spectrum), but that should be fixed now. Just added a missing unit test and now it should be ready for review.

File: tests/test_spectrum.py [hunk 1]
```diff
@@ -26,6 +26,69 @@ def _create_test_spectrum_with_intensities(intensities):
...
+    peaks_mz += 100.0
+    assert numpy.all(spectrum.peaks.mz == numpy.array([100.0, 101.0])), "Expected different peaks.mz"
+    metadata = spectrum.metadata
+    metadata["added_info"] = "this"
+    assert spectrum.metadata == {'testdata': 1}, "Expected metadata to remain unchanged"
+
+
+def test_comparing_spectra_with_metadata():
+    """Test if spectra can be compared that contain numpy.arrays in the metadata.
+    (Failed in an earlier version)"""
```
Comment (florian-huber) [inline tests/test_spectrum.py]:
```suggestion
    """Test if spectra with (slightly) different metadata are correctly compared."""
```

File: integration-tests/test_user_workflow.py [hunk 2]
```diff
@@ -44,23 +45,31 @@ def apply_my_filters(s):
...
+        (references[48], queries[46], numpy.array([(0.9990141860269471, 27)], dtype=score_datatype)[0]),
+        (references[46], queries[50], numpy.array([(0.9988793406908721, 22)], dtype=score_datatype)[0]),
+        (references[50], queries[46], numpy.array([(0.9988793406908721, 22)], dtype=score_datatype)[0]),
+        (references[57], queries[59], numpy.array([(0.9982171275552503, 46)], dtype=score_datatype)[0]),
+        (references[59], queries[57], numpy.array([(0.9982171275552503, 46)], dtype=score_datatype)[0]),
+        (references[73], queries[74], numpy.array([(0.9973823244169199, 23)], dtype=score_datatype)[0]),
+        (references[74], queries[73], numpy.array([(0.9973823244169199, 23)], dtype=score_datatype)[0]),
     ]
-    assert actual_top10 == expected_top10
+    assert [x[2][0] for x in actual_top10] == pytest.approx([x[2][0] for x in expected_top10], 1e-8), \
```
Comment (sverhoeven) [inline integration-tests/test_user_workflow.py:68]: Sad, that the asserts now must be done per column, making it harder to read. 
Looked at [numpy.testing.assert_allclose](https://numpy.org/devdocs/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose), but couldn't get it to work.

For posterity:
```python
    assert [x[0:2] for x in actual_top10] == [x[0:2] for x in expected_top10], \
        "Expected different references and queries."
    assert numpy.testing.assert_allclose(
        numpy.array([x[2] for x in actual_top10]),
        numpy.array([x[2] for x in expected_top10])
    ), "Expected different scores."
```

File: matchms/Scores.py [hunk 3]
```diff
@@ -39,9 +39,9 @@ class Scores:
         similarity_measure = CosineGreedy()
         scores = calculate_scores(references, queries, similarity_measure)
 
-        for (reference, query, score, n_matching) in scores:
+        for (reference, query, score) in scores:
             print(f"Cosine score between {reference.get('id')} and {query.get('id')}" +
-                  f" is {score:.2f} with {n_matching} matched peaks")
+                  f" is {score['score']:.2f} with {score['matches']} matched peaks")
 
     Should output
```
Comment (sverhoeven) [inline matchms/Scores.py:47]: Line 95-97 should be updated, as there is no longer a similarity function which returns a tuple.

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,6 +7,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+
+- scores_by_query and scores_by reference now accept sort=True to return sorted scores [#153](https://github.com/matchms/matchms/pull/153)
+
+### Changed
+
+- Scores.scores is now returning a structured array [#153](https://github.com/matchms/matchms/pull/153)
```
Comment (sverhoeven) [inline CHANGELOG.md:16]: Please also mention the `BaseSimilarity.sort` and `BaseSimilarity.score_datatype`, someone making a Spec2vec like library should be aware of the changes to the BaseSimilarity class.

File: matchms/Scores.py [hunk 5]
```diff
@@ -128,39 +128,46 @@ def calculate(self) -> Scores:
...
+                            sort: bool = False) -> numpy.ndarray:
         """Return all scores (not sorted) for the given reference spectrum.
 
         Parameters
         ----------
         reference
             Single reference Spectrum.
+        sort
+            Set to True to obtain the scores in a sorted way (relying on the
+            sorting function from the given similarity_function).
```
Comment (sverhoeven) [inline matchms/Scores.py]: Use correct name of function and link it
```suggestion
            :meth:`~.BaseSimilarity.sort` function from the given similarity_function).
```

File: matchms/Scores.py [hunk 6]
```diff
@@ -128,39 +128,46 @@ def calculate(self) -> Scores:
...
         """Return all scores (not sorted) for the given query spectrum.
 
         Parameters
         ----------
         query
             Single query Spectrum.
-
+        sort
+            Set to True to obtain the scores in a sorted way (relying on the
+            sorting function from the given similarity_function).
```
Comment (sverhoeven) [inline matchms/Scores.py]: Use correct name of function and link it
```suggestion
            :meth:`~.BaseSimilarity.sort` function from the given similarity_function).
```

File: matchms/Scores.py [hunk 7]
```diff
@@ -128,39 +128,46 @@ def calculate(self) -> Scores:
...
-            else:
-                selected_scores.append((self.queries[i], score))
+        if sort:
+            query_idx_sorted = self.similarity_function.sort(self._scores[selected_idx, :])
+            selected_scores = list(zip(self.queries[query_idx_sorted],
+                                       self._scores[selected_idx, query_idx_sorted].copy()))
+        else:
+            selected_scores = list(zip(self.queries, self._scores[selected_idx, :].copy()))
+
         return selected_scores
```
Comment (sverhoeven) [inline matchms/Scores.py]: Nice to see functional rewrite. Could be simplified to
```suggestion
            query_idx_sorted = self.similarity_function.sort(self._scores[selected_idx, :])
            return list(zip(self.queries[query_idx_sorted],
                                       self._scores[selected_idx, query_idx_sorted].copy()))
        return list(zip(self.queries, self._scores[selected_idx, :].copy()))
```

File: matchms/Scores.py [hunk 8]
```diff
@@ -192,12 +198,12 @@ def scores_by_query(self, query: QueriesType) -> numpy.ndarray:
...
-                selected_scores.append((self.references[i], *score))
-            else:
-                selected_scores.append((self.references[i], score))
+        if sort:
+            references_idx_sorted = self.similarity_function.sort(self._scores[:, selected_idx])
+            selected_scores = list(zip(self.references[references_idx_sorted],
+                                       self._scores[references_idx_sorted, selected_idx].copy()))
+        else:
+            selected_scores = list(zip(self.references, self._scores[:, selected_idx].copy()))
         return selected_scores
```
Comment (sverhoeven) [inline matchms/Scores.py]: Could be simplified to
```suggestion
            return list(zip(self.references[references_idx_sorted],
                                       self._scores[references_idx_sorted, selected_idx].copy()))
        return list(zip(self.references, self._scores[:, selected_idx].copy()))
```

File: matchms/similarity/BaseSimilarity.py [hunk 9]
```diff
@@ -59,3 +62,21 @@ def matrix(self, references: List[SpectrumType], queries: List[SpectrumType],
...
+            1D Array of scores.
+
+        Returns
+        -------
+        idx_sorted
+            Indexes of sorted scores.
+        """
+        if scores.dtype.names is None:
+            return scores.argsort()[::-1]
+        return scores["score"].argsort()[::-1]
```
Comment (sverhoeven) [inline matchms/similarity/BaseSimilarity.py]: Don't think the base class should look at the incoming dtype. It should sort according to the score_datatype. So it should only consist out of line 81. Line 82 is specific to the Cosine classes it should not be part of the base.

Classes which override score_datatype should have their own `sort()` if needed. The `sort() could be implemented in an intermediate abstract class or in the class itself.

File: readthedocs/index.rst [hunk 10]
```diff
@@ -74,14 +74,14 @@ Below is a small example of using matchms to calculate the Cosine score between
...
+        (reference, query, score) = score
         # Ignore scores between same spectrum and
         # pairs which have less than 20 peaks in common
-        if reference is not query and n_matching >= 20:
+        if reference is not query and score['matches'] >= 20:
             print(f"Reference scan id: {reference.metadata['scans']}")
             print(f"Query scan id: {query.metadata['scans']}")
-            print(f"Score: {score:.4f}")
-            print(f"Number of matching peaks: {n_matching}")
+            print(f"Score: {score['score']:.4f}")
```
Comment (sverhoeven) [inline readthedocs/index.rst:83]: https://github.com/matchms/matchms/blob/add_querying_options/README.rst#example should also use `score['score']` instead tuple splat.

**Review by sverhoeven — CHANGES_REQUESTED** [review]
Overall looks good, docstrings render OK, additional tests result in nice coverage.
I like the approach of having a `BaseSimilarity.sort()` method. (I might have suggested it before, but good to see that it was also implementable in a unobtrusive way)

Calculation scores between 2 spectra is failing to return a structured array

I tried
```ipython
In [1]:         import numpy as np
   ...:         from matchms import calculate_scores
   ...:         from matchms import Spectrum
   ...:         from matchms.similarity import CosineGreedy
   ...: 
   ...:         spectrum_1 = Spectrum(mz=np.array([100, 150, 200.]),
   ...:                               intensities=np.array([0.7, 0.2, 0.1]),
   ...:                               metadata={'id': 'spectrum1'})
   ...:         spectrum_2 = Spectrum(mz=np.array([100, 140, 190.]),
   ...:                               intensities=np.array([0.4, 0.2, 0.1]),
   ...:                               metadata={'id': 'spectrum2'})

In [2]: similarity_measure = CosineGreedy()

In [3]: scores = calculate_scores([spectrum_1],[spectrum_2], similarity_measure)

In [6]: scores._scores
Out[6]: array([[(0.831479419283098, 1)]], dtype=object)

In [7]: list(scores)
Out[7]: 
[(<matchms.Spectrum.Spectrum at 0x7fb29ac050d0>,
  <matchms.Spectrum.Spectrum at 0x7fb29ac05070>,
  0.831479419283098,
  1)]

```

Instead of `0.831479419283098,  1` I expected to get back a dictionary or structured array.
Can you add a test for this use case?

File: integration-tests/test_user_workflow.py (hunk 2 above)
Comment (florian-huber) [reply integration-tests/test_user_workflow.py:68]: Yes, I also wasted some time on this :D ...

File: matchms/Scores.py (hunk 7 above)
Comment (florian-huber) [reply matchms/Scores.py]: Nice, thanks, looks much better this way (somehow I keep overlooking those things a bit...).

File: tests/test_scores.py [hunk 11]
```diff
@@ -2,36 +2,41 @@
...
+    score_datatype = [("score", numpy.unicode_, 16), ("len", numpy.int32)]
+
     def __init__(self):
         """constructor"""
 
     def pair(self, reference, query):
         """necessary pair computation method"""
         s = reference + query
-        return s, len(s)
+        return numpy.array([(s, len(s))], dtype=self.score_datatype)
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply tests/test_scores.py:21]: That is indeed more consistent. I added this to our scoring functions.

File: matchms/Scores.py [hunk 12]
```diff
@@ -178,10 +185,9 @@ def scores_by_query(self, query: QueriesType) -> numpy.ndarray:
             references = [spectrum_1, spectrum_2, spectrum_3]
             queries = [spectrum_2, spectrum_3, spectrum_4]
 
-            scores = Scores(references, queries, CosineGreedy()).calculate()
-            selected_scores = scores.scores_by_query(spectrum_4)
-            selected_scores.sort(key=lambda s: s[1], reverse=True)
-            print([np.round(x[1], 3) for x in selected_scores])
+            scores = calculate_scores(references, queries, CosineGreedy())
+            selected_scores = scores.scores_by_query(spectrum_4, sort=True)
+            print([x[1]["score"].round(3) for x in selected_scores])
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/Scores.py:187]: Great, thanks!

File: readthedocs/index.rst (hunk 10 above)
Comment (florian-huber) [reply readthedocs/index.rst:83]: Good catch. I changed the example.

File: tests/test_scores.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (florian-huber) [reply tests/test_scores.py:21]: Thanks Stefan. I now implemented the structured arrays for the pair methods as well.

File: matchms/similarity/BaseSimilarity.py (hunk 9 above)
Comment (florian-huber) [reply matchms/similarity/BaseSimilarity.py]: I reduced it to line 81 as suggested. In fact, that even works for all scores we implemented so far (without need to define own sort() method).

**Comment by florian-huber** [comment]
Thanks a lot for the reviewing Stefan, that was very helpful! I believe I could address your comments and will merge this PR now.

**Pull request merged**